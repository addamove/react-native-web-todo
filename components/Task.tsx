import { useNavigation } from "@react-navigation/native";
import React from "react";
import { GestureResponderEvent, StyleSheet } from "react-native";
import { Icon, ListItem } from "react-native-elements";
import { TouchableOpacity } from "react-native-gesture-handler";

export interface ITask {
  id: string;
  title: string;
  description?: string;
  status: "done" | "todo" | "archived";
}
export type ITaskProps = ITask & {
  onCheckboxPress: (event: GestureResponderEvent) => void;
  onDeleteItem: (id: string) => void;
};

export const Task: React.VFC<ITaskProps> = ({
  status,
  title,
  onDeleteItem,
  onCheckboxPress,
  id,
}) => {
  const navigation = useNavigation();

  return (
    <ListItem containerStyle={styles.container} bottomDivider>
      <ListItem.CheckBox
        checkedColor="#794BC4"
        checked={status === "done"}
        onPress={onCheckboxPress}
      />
      <ListItem.Content>
        <ListItem.Title>{title}</ListItem.Title>
      </ListItem.Content>
      <TouchableOpacity
        accessible
        accessibilityRole="button"
        testID="delete-button"
        accessibilityLabel="Delete Item"
        onPress={() => onDeleteItem(id)}
      >
        <Icon
          style={{ marginRight: "0.5rem" }}
          name="trash"
          type="font-awesome"
          color="red"
          size={24}
        />
      </TouchableOpacity>
      <ListItem.Chevron
        testID="navigator"
        onPress={() => navigation.navigate("Item", { id })}
        size={24}
      />
    </ListItem>
  );
};

const styles = StyleSheet.create({
  container: {
    width: "100%",
  },
  divider: {
    width: 1,
  },
});
