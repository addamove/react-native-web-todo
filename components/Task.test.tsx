import React from "react";
import { fireEvent, render } from "@testing-library/react-native";
import { ITaskProps, Task } from "./Task";

const navigateJestMockFn = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useFocusEffect: () => jest.fn(),
    useNavigation: () => ({
      navigate: navigateJestMockFn,
    }),
  };
});

describe("Task tests", () => {
  const taskProps: ITaskProps = {
    onDeleteItem: jest.fn(),
    onCheckboxPress: jest.fn(),
    description: "description",
    id: "123",
    status: "todo",
    title: "title",
  };

  it("Toggle status", () => {
    const { getByTestId } = render(<Task {...taskProps} />);
    const checkbox = getByTestId("checkbox");
    fireEvent.press(checkbox);
    fireEvent.press(checkbox);
    expect(taskProps.onCheckboxPress).toBeCalledTimes(2);
  });
  it("Fire Deliting", () => {
    const { getAllByA11yLabel, getByTestId, toJSON, debug } = render(
      <Task {...taskProps} />
    );
    const deleteButton = getByTestId("delete-button");
    fireEvent.press(deleteButton);
    expect(taskProps.onDeleteItem).toBeCalledTimes(1);
  });
  it("Fire navigate", () => {
    const { getByTestId } = render(<Task {...taskProps} />);
    const navigateButton = getByTestId("navigator");
    fireEvent.press(navigateButton);
    expect(navigateJestMockFn).toBeCalledTimes(1);
  });

  it("Should render snapshot", () => {
    const { toJSON } = render(<Task {...taskProps} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
