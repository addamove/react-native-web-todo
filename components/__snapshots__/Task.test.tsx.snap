// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task tests Should render snapshot 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderBottomWidth": 0.5,
        "borderColor": "#bcbbc1",
        "flexDirection": "row",
        "padding": 14,
        "width": "100%",
      }
    }
    testID="padView"
  >
    <View
      accessibilityRole="checkbox"
      accessibilityState={
        Object {
          "checked": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "margin": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "opacity": 1,
          "padding": 0,
        }
      }
      testID="checkbox"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#bfbfbf",
                "fontSize": 24,
              },
              Object {
                "minWidth": 24,
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "paddingLeft": 16,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flex": 1,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
      theme={
        Object {
          "colors": Object {
            "black": "#242424",
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "grey": "rgba(0, 0, 0, 0.54)",
                "primary": "#2196f3",
                "searchBg": "#dcdce1",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "default": Object {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "ios": Object {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "web": Object {
                "error": "#ff190c",
                "grey": "#393e42",
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#ca71eb",
            "success": "#52c41a",
            "warning": "#faad14",
            "white": "#ffffff",
          },
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "transparent",
            "color": "#242424",
            "fontSize": 17,
          }
        }
        testID="listItemTitle"
      >
        title
      </Text>
    </View>
    <View
      style={
        Object {
          "paddingLeft": 16,
        }
      }
    />
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
      testID="delete-button"
    >
      <View
        accessibilityLabel="Delete Item"
        accessibilityRole="button"
        accessible={true}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Object {
                "marginRight": "0.5rem",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "red",
                      "fontSize": 24,
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
                testID="iconIcon"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RNGestureHandlerButton>
    <View
      style={
        Object {
          "paddingLeft": 16,
        }
      }
    />
    <View
      style={
        Object {
          "alignSelf": "center",
          "overflow": "hidden",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="navigator"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#D1D1D6",
                  "fontSize": 24,
                },
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "fontFamily": "ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="iconIcon"
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
